In [1]:
import pandas as pd
import numpy as np
%config Completer.use_jedi = False

In [19]:
a=pd.date_range(start="20210101",end="20210201")
b=pd.date_range(start="20210115",end="20210215")

c=b-a
a,b

(DatetimeIndex(['2021-01-01', '2021-01-02', '2021-01-03', '2021-01-04',
                '2021-01-05', '2021-01-06', '2021-01-07', '2021-01-08',
                '2021-01-09', '2021-01-10', '2021-01-11', '2021-01-12',
                '2021-01-13', '2021-01-14', '2021-01-15', '2021-01-16',
                '2021-01-17', '2021-01-18', '2021-01-19', '2021-01-20',
                '2021-01-21', '2021-01-22', '2021-01-23', '2021-01-24',
                '2021-01-25', '2021-01-26', '2021-01-27', '2021-01-28',
                '2021-01-29', '2021-01-30', '2021-01-31', '2021-02-01'],
               dtype='datetime64[ns]', freq='D'),
 DatetimeIndex(['2021-01-15', '2021-01-16', '2021-01-17', '2021-01-18',
                '2021-01-19', '2021-01-20', '2021-01-21', '2021-01-22',
                '2021-01-23', '2021-01-24', '2021-01-25', '2021-01-26',
                '2021-01-27', '2021-01-28', '2021-01-29', '2021-01-30',
                '2021-01-31', '2021-02-01', '2021-02-02', '2021-02-03',
             

In [20]:
c

TimedeltaIndex(['14 days', '14 days', '14 days', '14 days', '14 days',
                '14 days', '14 days', '14 days', '14 days', '14 days',
                '14 days', '14 days', '14 days', '14 days', '14 days',
                '14 days', '14 days', '14 days', '14 days', '14 days',
                '14 days', '14 days', '14 days', '14 days', '14 days',
                '14 days', '14 days', '14 days', '14 days', '14 days',
                '14 days', '14 days'],
               dtype='timedelta64[ns]', freq=None)

In [21]:
c.days

Int64Index([14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14,
            14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 14],
           dtype='int64')

## 1-2588

https://www.acmicpc.net/problem/2588

In [93]:
# A,B=list(map(int, input().split()))
A=int(input())
B=int(input())

print(A*int(str(B)[2]))
print(A*int(str(B)[1]))
print(A*int(str(B)[0]))
print(A*B)

 472
 385


2360
3776
1416
181720


## 6단계 함수

### 6.1 정수 N개의 합

In [89]:
def solve(a:list) -> int:
#     if isinstance(a, list)==False:
#         raise TypeError("not list")
    return sum(a)

solve([1,2])

3

### 6.2 셀프 넘버

셀프 넘버는 1949년 인도 수학자 D.R. Kaprekar가 이름 붙였다. 양의 정수 n에 대해서 d(n)을 n과 n의 각 자리수를 더하는 함수라고 정의하자. 예를 들어, d(75) = 75+7+5 = 87이다.

양의 정수 n이 주어졌을 때, 이 수를 시작해서 n, d(n), d(d(n)), d(d(d(n))), ...과 같은 무한 수열을 만들 수 있다. 

예를 들어, 33으로 시작한다면 다음 수는 33 + 3 + 3 = 39이고, 그 다음 수는 39 + 3 + 9 = 51, 다음 수는 51 + 5 + 1 = 57이다. 이런식으로 다음과 같은 수열을 만들 수 있다.

33, 39, 51, 57, 69, 84, 96, 111, 114, 120, 123, 129, 141, ...

n을 d(n)의 생성자라고 한다. 위의 수열에서 33은 39의 생성자이고, 39는 51의 생성자, 51은 57의 생성자이다. 생성자가 한 개보다 많은 경우도 있다. 예를 들어, 101은 생성자가 2개(91과 100) 있다. 

생성자가 없는 숫자를 셀프 넘버라고 한다. 100보다 작은 셀프 넘버는 총 13개가 있다. 1, 3, 5, 7, 9, 20, 31, 42, 53, 64, 75, 86, 97

10000보다 작거나 같은 셀프 넘버를 한 줄에 하나씩 출력하는 프로그램을 작성하시오.

In [112]:
def d(number):
    return sum([int(n) for n in str(number)])+number

nums=[d(i) for i in range(1,10001)]
num_line=[i for i in range(1,10001)]
for i in sorted(list(set(num_line)-set(nums))):
    print(i)

[1,
 3,
 5,
 7,
 9,
 20,
 31,
 42,
 53,
 64,
 75,
 86,
 97,
 108,
 110,
 121,
 132,
 143,
 154,
 165,
 176,
 187,
 198,
 209,
 211,
 222,
 233,
 244,
 255,
 266,
 277,
 288,
 299,
 310,
 312,
 323,
 334,
 345,
 356,
 367,
 378,
 389,
 400,
 411,
 413,
 424,
 435,
 446,
 457,
 468,
 479,
 490,
 501,
 512,
 514,
 525,
 536,
 547,
 558,
 569,
 580,
 591,
 602,
 613,
 615,
 626,
 637,
 648,
 659,
 670,
 681,
 692,
 703,
 714,
 716,
 727,
 738,
 749,
 760,
 771,
 782,
 793,
 804,
 815,
 817,
 828,
 839,
 850,
 861,
 872,
 883,
 894,
 905,
 916,
 918,
 929,
 940,
 951,
 962,
 973,
 984,
 995,
 1006,
 1021,
 1032,
 1043,
 1054,
 1065,
 1076,
 1087,
 1098,
 1109,
 1111,
 1122,
 1133,
 1144,
 1155,
 1166,
 1177,
 1188,
 1199,
 1210,
 1212,
 1223,
 1234,
 1245,
 1256,
 1267,
 1278,
 1289,
 1300,
 1311,
 1313,
 1324,
 1335,
 1346,
 1357,
 1368,
 1379,
 1390,
 1401,
 1412,
 1414,
 1425,
 1436,
 1447,
 1458,
 1469,
 1480,
 1491,
 1502,
 1513,
 1515,
 1526,
 1537,
 1548,
 1559,
 1570,
 1581,
 1592,


### 6.3 한수

In [62]:
N=int(input())
def arithmetic(number):
    cnt=0
    for i in range(1,number+1):
        if len(str(i)) in [1,2]:
            cnt+=1
        elif len(set([a-b for a,b in zip([int(j) for j in str(i)[1:]], [int(j) for j in str(i)[:-1]])]))==1:
            cnt+=1
    return cnt
print(arithmetic(N))

 210


In [49]:
number=101
print([int(j) for j in str(number)], [int(j) for j in str(number)])
print([int(j) for j in str(number)[1:]], [int(j) for j in str(number)[:-1]])
print([a-b for a,b in zip([int(j) for j in str(number)[1:]], [int(j) for j in str(number)[:-1]])])
print(len(set([a-b for a,b in zip([int(j) for j in str(number)[1:]], [int(j) for j in str(number)[:-1]])])))

[1, 0, 1] [1, 0, 1]
[0, 1] [1, 0]
[-1, 1]
2


## 7단계 문자열

### 7.1 아스키코드
- 알파벳 소문자, 대문자, 숫자 0-9중 하나가 주어졌을 때, 주어진 글자의 아스키 코드값을 출력하는 프로그램을 작성하시오.

In [120]:
import re
char=input()
regex=r"[a-zA-z0-9]"
p=re.compile(regex)
if isinstance(p.match(char).group(), type(char)):
    print(ord(char))
else:
    print("Error")

 A


65


### 7.2 숫자의 합

- 문제 : N개의 숫자가 공백 없이 쓰여있다. 이 숫자를 모두 합해서 출력하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 숫자의 개수 N (1 ≤ N ≤ 100)이 주어진다. 둘째 줄에 숫자 N개가 공백없이 주어진다.

- 출력 : 입력으로 주어진 숫자 N개의 합을 출력한다.

In [159]:
# 이게 정답인데 문제가 질이 안 좋다
import random
A=int(input())
B=input()
print(sum(int(i) for i in B))

 5
 12345


15


In [152]:
# 첫 번째 input만큼의 숫자를 가지는 string을 B변수에 정의한 후에 그것의 합을 구하게 하는 것이 나을 듯
import random
l=[]
for _ in range(int(input())):
    l.append(str(random.randint(0,9)))
N="".join(l)
print(N)
print(sum([int(i) for i in N]))

 10


1596592609
52


### 7.3 알파벳 찾기

In [167]:
list(set(["b","a","e","e","j"]))

['b', 'j', 'a', 'e']

In [241]:
from collections import Counter
inp=list("baekjoon")
# count=Counter(inp)
# over=[inp[idx] for idx, i in enumerate(list(count.values())) if i > 1]

# over_cnt=count[over[0]]
# for i in range(2,over_cnt+1):
#     inp[i]=" "


try:
    inp.index("1")
except:
    l.append(-1)

ValueError: '1' is not in list

In [246]:
list(map(str, l))

['1',
 '0',
 '-1',
 '-1',
 '2',
 '-1',
 '-1',
 '-1',
 '-1',
 '4',
 '3',
 '-1',
 '-1',
 '7',
 '5',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1',
 '-1']

In [258]:
letters="abcdefghijklmnopqrstuvwxyz"
string=input()
for l in letters:
    print(string.find(l), end= ' ')

 baekjoon


1 0 -1 -1 2 -1 -1 -1 -1 4 3 -1 -1 7 5 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 

In [260]:
inp=list(input())
l=[]
for idx, a in enumerate("abcdefghijklmnopqrstuvwxyz"):
    if a in inp:
        l.append(inp.index(a))
    else:
        l.append(-1)
print(" ".join(list(map(str, l))))

 baekjoon


1 0 -1 -1 2 -1 -1 -1 -1 4 3 -1 -1 7 5 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1


### 7.4 문자열 반복

- 문제 : 문자열 S를 입력받은 후에, 각 문자를 R번 반복해 새 문자열 P를 만든 후 출력하는 프로그램을 작성하시오. 즉, 첫 번째 문자를 R번 반복하고, 두 번째 문자를 R번 반복하는 식으로 P를 만들면 된다. S에는 QR Code "alphanumeric" 문자만 들어있다.
    - QR Code "alphanumeric" 문자는 0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ\$%*+-./: 이다.

<br>

- 입력 : 첫째 줄에 테스트 케이스의 개수 T(1 ≤ T ≤ 1,000)가 주어진다. 각 테스트 케이스는 반복 횟수 R(1 ≤ R ≤ 8), 문자열 S가 공백으로 구분되어 주어진다. S의 길이는 적어도 1이며, 20글자를 넘지 않는다. 

- 출력 : 각 테스트 케이스에 대해 P를 출력한다.

In [305]:
for _ in range(int(input())):R, S = list(input().split());print( "".join([s*int(R) for s in S]) )

 3
 1 A


A


 2 BD


BBDD


 3 BDF


BBBDDDFFF


20

### 7.5 단어공부 

- 문제 : 알파벳 대소문자로 된 단어가 주어지면, 이 단어에서 가장 많이 사용된 알파벳이 무엇인지 알아내는 프로그램을 작성하시오. 단, 대문자와 소문자를 구분하지 않는다.

- 입력 : 첫째 줄에 알파벳 대소문자로 이루어진 단어가 주어진다. 주어지는 단어의 길이는 1,000,000을 넘지 않는다.

- 출력 : 첫째 줄에 이 단어에서 가장 많이 사용된 알파벳을 대문자로 출력한다. 단, 가장 많이 사용된 알파벳이 여러 개 존재하는 경우에는 ?를 출력한다.

In [357]:
from collections import Counter
dic=Counter(input().lower())
l=[k.upper() for k, v in dic.items() if v == max(dic.values())]
print("?" if len(l)>1 else l[0])

 aaaa231aabcc


A


In [360]:
"ASDW".count("S")

1

### 7.6 단어의 개수
- 문제 : 영어 대소문자와 띄어쓰기만으로 이루어진 문자열이 주어진다. 이 문자열에는 몇 개의 단어가 있을까? 이를 구하는 프로그램을 작성하시오. 단, 한 단어가 여러 번 등장하면 등장한 횟수만큼 모두 세어야 한다.

- 입력 : 첫 줄에 영어 대소문자와 띄어쓰기로 이루어진 문자열이 주어진다. 이 문자열의 길이는 1,000,000을 넘지 않는다. 단어는 띄어쓰기 한 개로 구분되며, 공백이 연속해서 나오는 경우는 없다. 또한 문자열의 앞과 뒤에는 공백이 있을 수도 있다.

- 출력 : 첫째 줄에 단어의 개수를 출력한다.

In [367]:
print(len(input().split()))

 asdfae asdfaew vbgf


3


### 7.7 상수

- 문제 : 상근이의 동생 상수는 수학을 정말 못한다. 상수는 숫자를 읽는데 문제가 있다. 이렇게 수학을 못하는 상수를 위해서 상근이는 수의 크기를 비교하는 문제를 내주었다. 상근이는 세 자리 수 두 개를 칠판에 써주었다. 그 다음에 크기가 큰 수를 말해보라고 했다.상수는 수를 다른 사람과 다르게 거꾸로 읽는다. 예를 들어, 734와 893을 칠판에 적었다면, 상수는 이 수를 437과 398로 읽는다. 따라서, 상수는 두 수중 큰 수인 437을 큰 수라고 말할 것이다.두 수가 주어졌을 때, 상수의 대답을 출력하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 상근이가 칠판에 적은 두 수 A와 B가 주어진다. 두 수는 같지 않은 세 자리 수이며, 0이 포함되어 있지 않다.

- 출력 : 첫째 줄에 상수의 대답을 출력한다.

In [396]:
# 수정 전
A,B=list(input().split())
A,B=int("".join([a for a in A][::-1])), int("".join([b for b in B][::-1]))
print(A if A>B else B)

 123 456


654


In [397]:
# 수정 후
A,B=list(input().split())
A,B=int(A[::-1]), int(B[::-1])
print(A if A>B else B)

 123 456


654


In [403]:
# 더 수정 후
print(max([ int(i[::-1]) for i in input().split() ]))

 123 456


654


### 7.8 다이얼

- 문제 : 상근이의 할머니는 아래 그림과 같이 오래된 다이얼 전화기를 사용한다. 전화를 걸고 싶은 번호가 있다면, 숫자를 하나를 누른 다음에 금속 핀이 있는 곳 까지 시계방향으로 돌려야 한다. 숫자를 하나 누르면 다이얼이 처음 위치로 돌아가고, 다음 숫자를 누르려면 다이얼을 처음 위치에서 다시 돌려야 한다. 숫자 1을 걸려면 총 2초가 필요하다. 1보다 큰 수를 거는데 걸리는 시간은 이보다 더 걸리며, 한 칸 옆에 있는 숫자를 걸기 위해선 1초씩 더 걸린다. 상근이의 할머니는 전화 번호를 각 숫자에 해당하는 문자로 외운다. 즉, 어떤 단어를 걸 때, 각 알파벳에 해당하는 숫자를 걸면 된다. 예를 들어, UNUCIC는 868242와 같다. 할머니가 외운 단어가 주어졌을 때, 이 전화를 걸기 위해서 필요한 최소 시간을 구하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 알파벳 대문자로 이루어진 단어가 주어진다. 단어의 길이는 2보다 크거나 같고, 15보다 작거나 같다.

- 출력 : 첫째 줄에 다이얼을 걸기 위해서 필요한 최소 시간을 출력한다.

In [443]:
total_time=0
t, w = [],[]
for i, asc in enumerate(range(1, 27),1):
    
    time=((i)//3.1)+3
    word=chr(asc+64)
    
    if word in ["Y","Z","S","V"]:
        time=time-1
    
    t.append(time)
    w.append(word)

for word in input():
    total_time += t[w.index(word)]
print(int(total_time))

 WA


13


In [448]:
{c:i for i,cs in enumerate(["ABC","DEF"],3) for c in cs}

{'A': 3, 'B': 3, 'C': 3, 'D': 4, 'E': 4, 'F': 4}

### 7.9 크로아티아 알파벳

- 문제 : 예전에는 운영체제에서 크로아티아 알파벳을 입력할 수가 없었다. 따라서, 다음과 같이 크로아티아 알파벳을 변경해서 입력했다.
|크로아티아|알파벳 변경|
|---|---|
|č|c=|
|ć|c-|
|dž|dz=|
|đ|d-|
|lj|lj|
|nj|nj|
|š|s=|
|ž|z=|
|||

- 예를 들어, ljes=njak은 크로아티아 알파벳 6개(lj, e, š, nj, a, k)로 이루어져 있다. 단어가 주어졌을 때, 몇 개의 크로아티아 알파벳으로 이루어져 있는지 출력한다. dž는 무조건 하나의 알파벳으로 쓰이고, d와 ž가 분리된 것으로 보지 않는다. lj와 nj도 마찬가지이다. 위 목록에 없는 알파벳은 한 글자씩 센다.

- 입력 : 첫째 줄에 최대 100글자의 단어가 주어진다. 알파벳 소문자와 '-', '='로만 이루어져 있다. 단어는 크로아티아 알파벳으로 이루어져 있다. 문제 설명의 표에 나와있는 알파벳은 변경된 형태로 입력된다.

- 출력 : 입력으로 주어진 단어가 몇 개의 크로아티아 알파벳으로 이루어져 있는지 출력한다.

In [465]:
inp=input()
print(len(inp) - (inp.count("=") + inp.count("-") + inp.count("nj") + inp.count("lj") + inp.count("dz=")))

 ljes=njak


6


In [483]:
# 다른 사람
c=input().count
print(c('')-1 - sum(map(c,["-","=","nj","lj","dz="])))

 ljes=njak


6


### 7.10 그룹 단어 체커

- 문제 : 그룹 단어란 단어에 존재하는 모든 문자에 대해서, 각 문자가 연속해서 나타나는 경우만을 말한다. 예를 들면, ccazzzzbb는 c, a, z, b가 모두 연속해서 나타나고, kin도 k, i, n이 연속해서 나타나기 때문에 그룹 단어이지만, aabbbccb는 b가 떨어져서 나타나기 때문에 그룹 단어가 아니다. 단어 N개를 입력으로 받아 그룹 단어의 개수를 출력하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 단어의 개수 N이 들어온다. N은 100보다 작거나 같은 자연수이다. 둘째 줄부터 N개의 줄에 단어가 들어온다. 단어는 알파벳 소문자로만 되어있고 중복되지 않으며, 길이는 최대 100이다.

- 출력 : 첫째 줄에 그룹 단어의 개수를 출력한다.

In [ ]:
group_word=0

for _ in range(int(input())):
    word=input()
    
    duplicated={i:word.count(i) for i in word if word.count(i)>1}.keys()
    if len(duplicated)>0:
        group_word_semi=0
        for d_word in duplicated:
            idxs=[idx for idx, s in enumerate(word) if s==d_word]
            search=word[min(idxs):max(idxs)+1]
            if search.count(d_word) != len(search):
                break
            else:
                group_word_semi+=1
            if group_word_semi==len(duplicated):
                group_word+=1
    else:
        group_word+=1
print(group_word)

In [569]:
cnt=0
for i in range(int(input())):
    word = input()
    print(word)
    print(list(word))
    print(sorted(word,key=word.find))
    cnt +=list(word) == sorted(word,key=word.find)
print(cnt)

 1
 aabbaabab


aabbaabab
['a', 'a', 'b', 'b', 'a', 'a', 'b', 'a', 'b']
['a', 'a', 'a', 'a', 'a', 'b', 'b', 'b', 'b']
0


## 8단계 기본 수학

### 8.1 손익분기점 

- 문제 : 월드전자는 노트북을 제조하고 판매하는 회사이다. 노트북 판매 대수에 상관없이 매년 임대료, 재산세, 보험료, 급여 등 A만원의 고정 비용이 들며, 한 대의 노트북을 생산하는 데에는 재료비와 인건비 등 총 B만원의 가변 비용이 든다고 한다. 예를 들어 A=1,000, B=70이라고 하자. 이 경우 노트북을 한 대 생산하는 데는 총 1,070만원이 들며, 열 대 생산하는 데는 총 1,700만원이 든다. 노트북 가격이 C만원으로 책정되었다고 한다. 일반적으로 생산 대수를 늘려 가다 보면 어느 순간 총 수입(판매비용)이 총 비용(=고정비용+가변비용)보다 많아지게 된다. 최초로 총 수입이 총 비용보다 많아져 이익이 발생하는 지점을 손익분기점(BREAK-EVEN POINT)이라고 한다. A, B, C가 주어졌을 때, 손익분기점을 구하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 A, B, C가 빈 칸을 사이에 두고 순서대로 주어진다. A, B, C는 21억 이하의 자연수이다.

- 출력 : 첫 번째 줄에 손익분기점 즉 최초로 이익이 발생하는 판매량을 출력한다. 손익분기점이 존재하지 않으면 -1을 출력한다.

- 타임초과된 코드

In [21]:
# A : 고정비용 1000
# B : 가변비용 70
# C : 노트북 가격 170

A,B,C = list(map(int, input().split()))

Total_A, Total_B, Total_C = A, B, C
cnt=0
while Total_C <= Total_A+Total_B:
    if B >= C:
        cnt=-1
        break
    # start
    if cnt == 0:
        Total_A = A
        Total_B = B
        Total_C = C
    else:
        Total_B += B
        Total_C += C
    cnt=cnt+1

print(cnt)

 2100 9 10


A+B: 2109 C: 10
A+B: 2118 C: 20
A+B: 2127 C: 30
A+B: 2136 C: 40
A+B: 2145 C: 50
A+B: 2154 C: 60
A+B: 2163 C: 70
A+B: 2172 C: 80
A+B: 2181 C: 90
A+B: 2190 C: 100
A+B: 2199 C: 110
A+B: 2208 C: 120
A+B: 2217 C: 130
A+B: 2226 C: 140
A+B: 2235 C: 150
A+B: 2244 C: 160
A+B: 2253 C: 170
A+B: 2262 C: 180
A+B: 2271 C: 190
A+B: 2280 C: 200
A+B: 2289 C: 210
A+B: 2298 C: 220
A+B: 2307 C: 230
A+B: 2316 C: 240
A+B: 2325 C: 250
A+B: 2334 C: 260
A+B: 2343 C: 270
A+B: 2352 C: 280
A+B: 2361 C: 290
A+B: 2370 C: 300
A+B: 2379 C: 310
A+B: 2388 C: 320
A+B: 2397 C: 330
A+B: 2406 C: 340
A+B: 2415 C: 350
A+B: 2424 C: 360
A+B: 2433 C: 370
A+B: 2442 C: 380
A+B: 2451 C: 390
A+B: 2460 C: 400
A+B: 2469 C: 410
A+B: 2478 C: 420
A+B: 2487 C: 430
A+B: 2496 C: 440
A+B: 2505 C: 450
A+B: 2514 C: 460
A+B: 2523 C: 470
A+B: 2532 C: 480
A+B: 2541 C: 490
A+B: 2550 C: 500
A+B: 2559 C: 510
A+B: 2568 C: 520
A+B: 2577 C: 530
A+B: 2586 C: 540
A+B: 2595 C: 550
A+B: 2604 C: 560
A+B: 2613 C: 570
A+B: 2622 C: 580
A+B: 2631 C: 590
A+B: 2

- 수정된 코드

In [57]:
A,B,C = list(map(int, input().split()))

if C-B < 0 :
    print(-1)
else :
    
    if int(str(A/(C-B)).split(".")[-1][0]) >= 5:
        print(round(A/(C-B)))
    else:
        print(round(A/C-B)+1)

 1000 40 170


1000
130
7.6923076923076925
90
8


- 다듬은 코드

In [66]:
A,B,C = map(int, input().split())

if C-B <= 0 :
    print(-1)
else :
    print(int(A/(C-B) + 1))

 1000 40 170


8


In [ ]:
# 다른 사람 코드
a,b,c = map(int, input().split())
if b >= c:
    print(-1)
else:
    print(a//(c-b)+1)

### 8.2 벌집
https://www.acmicpc.net/problem/2292
- 문제 : 위의 그림과 같이 육각형으로 이루어진 벌집이 있다. 그림에서 보는 바와 같이 중앙의 방 1부터 시작해서 이웃하는 방에 돌아가면서 1씩 증가하는 번호를 주소로 매길 수 있다. 숫자 N이 주어졌을 때, 벌집의 중앙 1에서 N번 방까지 최소 개수의 방을 지나서 갈 때 몇 개의 방을 지나가는지(시작과 끝을 포함하여)를 계산하는 프로그램을 작성하시오. 예를 들면, 13까지는 3개, 58까지는 5개를 지난다.

- 입력 : 첫째 줄에 N(1 ≤ N ≤ 1,000,000,000)이 주어진다.

- 출력 : 입력으로 주어진 방까지 최소 개수의 방을 지나서 갈 때 몇 개의 방을 지나는지 출력한다.


In [235]:
inp=int(input())
sm, step, n =1,0,0
while sm < inp:
    print("step:",step, "sum:",sm)
    step=step+6
    sm += step
    n+=1
print(n+1)

 58


step: 0 sum: 1
step: 6 sum: 7
step: 12 sum: 19
step: 18 sum: 37
5


In [ ]:
# 다른사람
print(int(-(-(3+(12*int(input())-3)**.5)//6)))

### 8.3 분수
https://www.acmicpc.net/problem/1193
- 문제 : 무한히 큰 배열에 다음과 같이 분수들이 적혀있다.

- 이와 같이 나열된 분수들을 1/1 -> 1/2 -> 2/1 -> 3/1 -> 2/2 -> … 과 같은 지그재그 순서로 차례대로 1번, 2번, 3번, 4번, 5번, … 분수라고 하자. X가 주어졌을 때, X번째 분수를 구하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 X(1 ≤ X ≤ 10,000,000)가 주어진다.

- 출력 : 첫째 줄에 분수를 출력한다.

In [239]:
# inp=int(input())
def get_step(number):
    step=0
    cnt=1
    while number>step:
        cnt+=1
        step+=cnt
        
    return step
inp=14
step=get_step(inp)
step

14

In [302]:
# inp=int(input())

def what_is(number):
    cumsum=0
    if number==1:
        return 1
    for step,_ in enumerate(range(10000001),1):
        print("step:",step)
        if cumsum<number and cumsum+step<number:
            cumsum+=step
        else:
            break
        
    print("cumsum:",cumsum)
    remaining=number-cumsum
    nums="".join(list(map(str, range(1,step+1))))
    if step%2==0:
        pass
    else:
        nums=nums[::-1]
    print(nums)
    print(remaining)
    return f"{nums[remaining-1]}/{nums[-remaining]}"

what_is(int(input()))

 6


step: 1
step: 2
step: 3
cumsum: 3
321
3


'1/3'

In [344]:
X=int(input())

def what_is(number):
    cumsum=0
    if number==1:
        return "1/1"
    for step,_ in enumerate(range(10000001),1):
        if cumsum<number and cumsum+step<number:
            cumsum+=step
        else:
            break
        
    remaining=number-cumsum
    nums="".join(list(map(str, range(1,step+1))))
    if step%2==0:
        pass
    else:
        nums=nums[::-1]
    return f"{nums[remaining-1]}/{nums[-remaining]}"

print(what_is(X))

 1


1/1


In [33]:
X=int(input())
cumsum=0
if X==1:
    print("1/1")
else:
    for step,_ in enumerate(range(1, 10000002)):
        if cumsum<X and cumsum+step<X:
            cumsum+=step
        else:
            break

    remaining=X-cumsum
    nums="".join(list(map(str, range(1,step+1))))
    if step%2==0:
        pass
    else:
        nums=nums[::-1]
    print(f"{nums[remaining-1]}/{nums[-remaining]}")

 20


5/2


### 8.4 달팽이는 올라가고 싶다

- 문제 : 땅 위에 달팽이가 있다. 이 달팽이는 높이가 V미터인 나무 막대를 올라갈 것이다. 달팽이는 낮에 A미터 올라갈 수 있다. 하지만, 밤에 잠을 자는 동안 B미터 미끄러진다. 또, 정상에 올라간 후에는 미끄러지지 않는다. 달팽이가 나무 막대를 모두 올라가려면, 며칠이 걸리는지 구하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 세 정수 A, B, V가 공백으로 구분되어서 주어진다. (1 ≤ B < A ≤ V ≤ 1,000,000,000)

- 출력 : 첫째 줄에 달팽이가 나무 막대를 모두 올라가는데 며칠이 걸리는지 출력한다.

In [74]:
A,B,V=map(int, input().split())
step=A-B
if (V-A)%step == 0:print((V-A)//step+1)    
elif (V-A)%step != 0:print((V-A)//step+2)

 2 1 5


4


### 8.5 ACM 호텔
https://www.acmicpc.net/problem/10250


In [46]:
T=int(input())

for _ in range(T):
    H,W,N=map(int, input().split())
    
    print(f"N:{N}, H:{H}, W:{W}")
    # 몫 + 1
    number=(N//(H+1))+1
    # 나머지
    floor=N%(H+1)
    print(f"가로 : {number}")
    print(f"세로 : {floor}")
#     if W_m==0:
#         W_m += 1
    # YYXX
#     X=W_m #4
#     Y=(H_m)
    print(f"floor:{floor}, Number:{number}")
    if floor == 0:
        number-=1
        floor=floor+1
#     if W_m>1:
#         H_m=H_m+1
#     if H_m > H:
#         H_m=1
#         W_m+=1
    print(f"floor:{floor}, Number:{number}")
    print(str(floor)+str(number).zfill(2))


 10
 8 4 1


N:1, H:8, W:4
가로 : 1
세로 : 1
floor:1, Number:1
floor:1, Number:1
101


 8 4 7


N:7, H:8, W:4
가로 : 1
세로 : 7
floor:7, Number:1
floor:7, Number:1
701


 8 4 8


N:8, H:8, W:4
가로 : 1
세로 : 8
floor:8, Number:1
floor:8, Number:1
801


 8 4 9


N:9, H:8, W:4
가로 : 2
세로 : 0
floor:0, Number:2
floor:1, Number:1
101


 8 4 10


N:10, H:8, W:4
가로 : 2
세로 : 1
floor:1, Number:2
floor:1, Number:2
102


 8 4 11


N:11, H:8, W:4
가로 : 2
세로 : 2
floor:2, Number:2
floor:2, Number:2
202


 8 4 15


N:15, H:8, W:4
가로 : 2
세로 : 6
floor:6, Number:2
floor:6, Number:2
602


 8 4 16


N:16, H:8, W:4
가로 : 2
세로 : 7
floor:7, Number:2
floor:7, Number:2
702


 8 4 17


N:17, H:8, W:4
가로 : 2
세로 : 8
floor:8, Number:2
floor:8, Number:2
802


KeyboardInterrupt: Interrupted by user

### 8.6 부녀회장이 될테야

- 문제 : 평소 반상회에 참석하는 것을 좋아하는 주희는 이번 기회에 부녀회장이 되고 싶어 각 층의 사람들을 불러 모아 반상회를 주최하려고 한다.이 아파트에 거주를 하려면 조건이 있는데, “a층의 b호에 살려면 자신의 아래(a-1)층의 1호부터 b호까지 사람들의 수의 합만큼 사람들을 데려와 살아야 한다” 는 계약 조항을 꼭 지키고 들어와야 한다. 아파트에 비어있는 집은 없고 모든 거주민들이 이 계약 조건을 지키고 왔다고 가정했을 때, 주어지는 양의 정수 k와 n에 대해 k층에 n호에는 몇 명이 살고 있는지 출력하라. 단, 아파트에는 0층부터 있고 각층에는 1호부터 있으며, 0층의 i호에는 i명이 산다.

- 입력 : 첫 번째 줄에 Test case의 수 T가 주어진다. 그리고 각각의 케이스마다 입력으로 첫 번째 줄에 정수 k, 두 번째 줄에 정수 n이 주어진다

- 출력 : 각각의 Test case에 대해서 해당 집에 거주민 수를 출력하라.

- 제한 : 1 ≤ k, n ≤ 14

In [56]:
k=2
n=12

In [57]:
_0=[i for i in range(1,15)]
_0

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]

In [58]:
_1=[sum(_0[:i]) for i in range(1,15)]
_1

[1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 66, 78, 91, 105]

In [59]:
_2=[sum(_1[:i]) for i in range(1,15)]
_2

[1, 4, 10, 20, 35, 56, 84, 120, 165, 220, 286, 364, 455, 560]

In [78]:
T=int(input())
# k=3
# n=3
for _ in range(T):
    k=int(input())
    n=int(input())
    under_floor=[i for i in range(1,15)]
    for i in range(1,k):
        _step=[sum(under_floor[:i]) for i in range(1,15)]
        under_floor=_step

    print(sum(under_floor[:n]))

 2
 1
 3


6


 2
 3


10


### 8.7 설탕 배달

- 문제 : 상근이는 요즘 설탕공장에서 설탕을 배달하고 있다. 상근이는 지금 사탕가게에 설탕을 정확하게 N킬로그램을 배달해야 한다. 설탕공장에서 만드는 설탕은 봉지에 담겨져 있다. 봉지는 3킬로그램 봉지와 5킬로그램 봉지가 있다. 상근이는 귀찮기 때문에, 최대한 적은 봉지를 들고 가려고 한다. 예를 들어, 18킬로그램 설탕을 배달해야 할 때, 3킬로그램 봉지 6개를 가져가도 되지만, 5킬로그램 3개와 3킬로그램 1개를 배달하면, 더 적은 개수의 봉지를 배달할 수 있다. 상근이가 설탕을 정확하게 N킬로그램 배달해야 할 때, 봉지 몇 개를 가져가면 되는지 그 수를 구하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 N이 주어진다. (3 ≤ N ≤ 5000)

- 출력 : 상근이가 배달하는 봉지의 최소 개수를 출력한다. 만약, 정확하게 N킬로그램을 만들 수 없다면 -1을 출력한다.

In [102]:
N=18

_5k = N//5
_3k = (N-_5k*5)//3

print(_5k, _3k)

if (N-_5k*5)%3 != 0:
    print(-1)
else:
    print(_5k + _3k)

3 1
4


In [135]:
N=17

_5k = N//5
_3k = N//3

for i in list(sorted(range(0,_5k+1), reverse=True)):
    print(f"i:{i}")
    print((_5k-i)*5, (N-(_5k-i)*5)%3)
    
    if (N-(_5k-i)*5)%3 == 0:
        
        print("AA",(_5k-i)+((N-i*5)//3))
        print((_5k-i), ((N-i*5)//3), f"N:{N}, N-i*5:{N-i*5}, ((N-i*5)//3):{((N-i*5)//3)}")
        break

# if (N-_5k*5)%3 != 0:
#     print(-1)
# else:
#     print(_5k + _3k)

i:3
0 2
i:2
5 0
AA 3
1 2 N:17, N-i*5:7, ((N-i*5)//3):2


In [194]:
N=int(input())
solution=None
for i in range(0,((N//5)+1)):
    
    # 3 2 1 0
    k5=((N-5*i)//5)*5
    
    print(k5, N-k5)
    if (N-k5)%3 == 0:
        solution=k5//5+(N-k5)//3
        print(solution)
        break
        
if solution==None:
    print(-1)

 4


0 4
-1


### 8.8 큰 수 A+B

- 문제 : 두 정수 A와 B를 입력받은 다음, A+B를 출력하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 A와 B가 주어진다. (0 < A,B < 1010000)

- 출력 : 첫째 줄에 A+B를 출력한다.

In [206]:
print(sum(map(int, input().split())))

 23456 23546


47002


### 8.9 Fly me to the Alpha Centauri

- 문제 : 우현이는 어린 시절, 지구 외의 다른 행성에서도 인류들이 살아갈 수 있는 미래가 오리라 믿었다. 그리고 그가 지구라는 세상에 발을 내려 놓은 지 23년이 지난 지금, 세계 최연소 ASNA 우주 비행사가 되어 새로운 세계에 발을 내려 놓는 영광의 순간을 기다리고 있다. 그가 탑승하게 될 우주선은 Alpha Centauri라는 새로운 인류의 보금자리를 개척하기 위한 대규모 생활 유지 시스템을 탑재하고 있기 때문에, 그 크기와 질량이 엄청난 이유로 최신기술력을 총 동원하여 개발한 공간이동 장치를 탑재하였다. 하지만 이 공간이동 장치는 이동 거리를 급격하게 늘릴 경우 기계에 심각한 결함이 발생하는 단점이 있어서, 이전 작동시기에 k광년을 이동하였을 때는 k-1 , k 혹은 k+1 광년만을 다시 이동할 수 있다. 예를 들어, 이 장치를 처음 작동시킬 경우 -1 , 0 , 1 광년을 이론상 이동할 수 있으나 사실상 음수 혹은 0 거리만큼의 이동은 의미가 없으므로 1 광년을 이동할 수 있으며, 그 다음에는 0 , 1 , 2 광년을 이동할 수 있는 것이다. ( 여기서 다시 2광년을 이동한다면 다음 시기엔 1, 2, 3 광년을 이동할 수 있다. ) 김우현은 공간이동 장치 작동시의 에너지 소모가 크다는 점을 잘 알고 있기 때문에 x지점에서 y지점을 향해 최소한의 작동 횟수로 이동하려 한다. 하지만 y지점에 도착해서도 공간 이동장치의 안전성을 위하여 y지점에 도착하기 바로 직전의 이동거리는 반드시 1광년으로 하려 한다. 김우현을 위해 x지점부터 정확히 y지점으로 이동하는데 필요한 공간 이동 장치 작동 횟수의 최솟값을 구하는 프로그램을 작성하라.

- 입력 : 입력의 첫 줄에는 테스트케이스의 개수 T가 주어진다. 각각의 테스트 케이스에 대해 현재 위치 x 와 목표 위치 y 가 정수로 주어지며, x는 항상 y보다 작은 값을 갖는다. (0 ≤ x < y < 231)

- 출력 : 각 테스트 케이스에 대해 x지점으로부터 y지점까지 정확히 도달하는데 필요한 최소한의 공간이동 장치 작동 횟수를 출력한다.

In [2]:
for i in range(int(input())):
    start, y = map(int, input().split())
    step=1
    cnt=0
    while start <= y-1:

        if start + step <= y-1:
            start = start + step
            step = step + 1
            cnt += 1
        else:
            break
    if y-start == 1:
        print(cnt + 1)
    else:
        print(cnt + 2)

 15
 0 15


6


 20 23


3


KeyboardInterrupt: Interrupted by user

In [2]:
import math 
for _ in range(int(input())):
    x,y=map(int, input().split())
    distance=y-x-2 if y-x != 1 else y-x-1
    print("distance", distance)
    step=int(math.ceil(distance/2))
    if y-x == 1:
        print(step+1)
    else:
        print(step+2)

KeyboardInterrupt: Interrupted by user

# 10단계 재귀

## 10.1 팩토리얼

- 문제 : 0보다 크거나 같은 정수 N이 주어진다. 이때, N!을 출력하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 정수 N(0 ≤ N ≤ 12)이 주어진다.

- 출력 : 첫째 줄에 N!을 출력한다.

In [17]:
a=1
for i in range(1,int(input())+1):
    a=a*i
print(a)

 0


1


In [18]:
a=1
print([a*i for i in range(1, int(input())+1 )])

 10


[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


## 10.2 피보나치 수 

- 문제 : 피보나치 수는 0과 1로 시작한다. 0번째 피보나치 수는 0이고, 1번째 피보나치 수는 1이다. 그 다음 2번째 부터는 바로 앞 두 피보나치 수의 합이 된다. 이를 식으로 써보면 Fn = Fn-1 + Fn-2 (n ≥ 2)가 된다. n=17일때 까지 피보나치 수를 써보면 다음과 같다. 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597 n이 주어졌을 때, n번째 피보나치 수를 구하는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 n이 주어진다. n은 20보다 작거나 같은 자연수 또는 0이다.

- 출력 : 첫째 줄에 n번째 피보나치 수를 출력한다.

In [1]:
def pibo(n):
    
    for i in range(n-1):
        if i == 0:
            fn_2=0
            fn_1=1
            
        else:
            fn_2=fn_1
            fn_1=fn
            
        fn=fn_1+fn_2
#         print(f"fn_2:{fn_2} fn_1:{fn_1} fn:{fn}")
    return fn

pibo(10)
    

55

## 11단계 브루스 포트

### 11.1 블랙잭

- 문제 : 카지노에서 제일 인기 있는 게임 블랙잭의 규칙은 상당히 쉽다. 카드의 합이 21을 넘지 않는 한도 내에서, 카드의 합을 최대한 크게 만드는 게임이다. 블랙잭은 카지노마다 다양한 규정이 있다. 한국 최고의 블랙잭 고수 김정인은 새로운 블랙잭 규칙을 만들어 상근, 창영이와 게임하려고 한다. 김정인 버전의 블랙잭에서 각 카드에는 양의 정수가 쓰여 있다. 그 다음, 딜러는 N장의 카드를 모두 숫자가 보이도록 바닥에 놓는다. 그런 후에 딜러는 숫자 M을 크게 외친다. 이제 플레이어는 제한된 시간 안에 N장의 카드 중에서 3장의 카드를 골라야 한다. 블랙잭 변형 게임이기 때문에, 플레이어가 고른 카드의 합은 M을 넘지 않으면서 M과 최대한 가깝게 만들어야 한다. N장의 카드에 써져 있는 숫자가 주어졌을 때, M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 구해 출력하시오.

- 입력 : 첫째 줄에 카드의 개수 N(3 ≤ N ≤ 100)과 M(10 ≤ M ≤ 300,000)이 주어진다. 둘째 줄에는 카드에 쓰여 있는 수가 주어지며, 이 값은 100,000을 넘지 않는 양의 정수이다. 합이 M을 넘지 않는 카드 3장을 찾을 수 있는 경우만 입력으로 주어진다.

- 출력 : 첫째 줄에 M을 넘지 않으면서 M에 최대한 가까운 카드 3장의 합을 출력한다.

- numpy 사용해서 블랙잭 구현

In [84]:
# 10 500
# 93 181 245 214 315 36 185 138 216 295

import itertools
import numpy as np

N, M=map(int, input().split())
C=list(map(int, input().split()))

result=list(itertools.combinations(C,3))
result=np.array(list(map(sum, result)))
idx=np.argwhere(result<=M)
print(result[idx][np.argmax(result[idx])][0])
# print(result[idx][np.argmax(result)])

 10 500
 93 181 245 214 315 36 185 138 216 295


497


In [98]:
import itertools

N, M=map(int, input().split())
C=list(map(int, input().split()))

result=list(itertools.combinations(C,3))
a=list(map(sum, result))
print(max([ a_ for a_ in a if a_ <= M]))

 5 21
 5 6 7 8 9


21


## 11.2 분해합

- 어떤 자연수 N이 있을 때, 그 자연수 N의 분해합은 N과 N을 이루는 각 자리수의 합을 의미한다. 어떤 자연수 M의 분해합이 N인 경우, M을 N의 생성자라 한다. 
- 예를 들어, 245의 분해합은 256(=245+2+4+5)이 된다. 따라서 245는 256의 생성자가 된다. 물론, 어떤 자연수의 경우에는 생성자가 없을 수도 있다. 
- 반대로, 생성자가 여러 개인 자연수도 있을 수 있다. 자연수 N이 주어졌을 때, N의 가장 작은 생성자를 구해내는 프로그램을 작성하시오.

- 입력 : 첫째 줄에 자연수 N(1 ≤ N ≤ 1,000,000)이 주어진다.

- 출력 : 첫째 줄에 답을 출력한다. 생성자가 없는 경우에는 0을 출력한다.

In [17]:

# 245 = 256의 생성자
# 256은 245의 분해합

256

## 11.3 덩치

In [40]:
# N=int(input())
xs,ys=[55,58,88,60,46],[185,183,186,175,155]
# for i in range(N):
#     x,y=map(int, input().split())
#     xs.append(x)
#     ys.append(y)

for x,y in zip(xs, ys):
    for x_,y_ in zip(xs, ys):
        grade=1
#         print(x,x_,y,y_)
        if x < x_ and y < y_:
            grade+=1
    print(grade)

1
1
1
1
1
